[JSC] Add megamorphic cache

Constellation opened the pull request:
#### 18d37d9fc2a08b00884449570d8d1011f0df2e95
<pre>
[JSC] Add megamorphic cache
<a href="https://bugs.webkit.org/show_bug.cgi?id=255060">https://bugs.webkit.org/show_bug.cgi?id=255060</a>
rdar://problem/107683366

Reviewed by Justin Michaud.

This patch adds megamorphic load cache finally, which exists in SpiderMonkey and V8. Our plan is adding megamorphic store cache next.

1. We move mayBePrototype bit from object cell to Structure. This allows us to disable IC setup for prototype objects by observing Structure.
   (And this is extremely rare). And every time we do some modification of the object, we can check `mayBePrototype()` and we can invalidate
   megamorphic cache integrity (VM::invalidateStructureChainIntegrity). This bit &amp; transition happens automatically when an object gets used as a prototype.
   But we mark all DOM prototypes and constructors a priori since they become prototype objects.
2. We integrate megamorphic load cache, which is 2-level hash table with 2 different hash functions. Primary table has 2048 entries and secondaryHash
   table has 512 entries, which is derived number from V8&apos;s megamorphic cache implementation. Our megamorphic cache&apos;s entry is StructureID + UniquedStringImpl.
   And the result is holder object and offset. Holder is null if property is missed. And holder is special marker (0x2) if it is own property.
   Every time objects used for prototype gets changed, whole megamorphic cache gets invalidated by incrementing epoch. This is surprisingly rare since
   we do not modify objects used for prototype after creating instances (And once the page gets stable state, we do not modify prototype objects).
   We use 2-level cache with 2 completely different hash functions to make it work well for various StructureID + UniquedStringImpl* pair.
3. We increment epoch of megamorphic cache when GC happens. So we do not keep entries valid over GC.

This change roughly offers 2-3x performance improvement for megamorphic access.

                                           ToT                     Patched

    megamorphic-own-load             99.3922+-1.9842     ^     41.2465+-0.0893        ^ definitely 2.4097x faster
    megamorphic-load                 14.8899+-0.0635     ^      5.0489+-0.0328        ^ definitely 2.9491x faster
    megamorphic-prototype-load      136.9894+-0.6358     ^     38.1426+-0.0790        ^ definitely 3.5915x faster
    megamorphic-miss                 98.3712+-1.6652     ^     31.9041+-0.0640        ^ definitely 3.0833x faster

* JSTests/microbenchmarks/megamorphic-miss.js: Added.
(f):
* JSTests/microbenchmarks/megamorphic-own-load.js: Added.
(f):
* JSTests/microbenchmarks/megamorphic-prototype-load.js: Added.
(f):
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/bytecode/AccessCase.cpp:
(JSC::AccessCase::create):
(JSC::AccessCase::guardedByStructureCheckSkippingConstantIdentifierCheck const):
(JSC::AccessCase::requiresIdentifierNameMatch const):
(JSC::AccessCase::requiresInt32PropertyCheck const):
(JSC::AccessCase::needsScratchFPR const):
(JSC::AccessCase::forEachDependentCell const):
(JSC::AccessCase::doesCalls const):
(JSC::AccessCase::canReplace const):
(JSC::AccessCase::runWithDowncast):
(JSC::AccessCase::canBeShared):
* Source/JavaScriptCore/bytecode/AccessCase.h:
* Source/JavaScriptCore/bytecode/CheckPrivateBrandStatus.cpp:
(JSC::CheckPrivateBrandStatus::CheckPrivateBrandStatus):
* Source/JavaScriptCore/bytecode/DeleteByStatus.cpp:
(JSC::DeleteByStatus::DeleteByStatus):
* Source/JavaScriptCore/bytecode/GetByStatus.cpp:
(JSC::GetByStatus::GetByStatus):
* Source/JavaScriptCore/bytecode/InByStatus.h:
* Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:
(JSC::InlineCacheCompiler::generateWithGuard):
(JSC::InlineCacheCompiler::generateImpl):
(JSC::InlineCacheCompiler::regenerate):
(WTF::printInternal):
* Source/JavaScriptCore/bytecode/InlineCacheCompiler.h:
(JSC::AccessGenerationResult::AccessGenerationResult):
(JSC::AccessGenerationResult::generatedMegamorphicCode const):
(JSC::AccessGenerationResult::generatedSomeCode const):
* Source/JavaScriptCore/bytecode/InstanceOfStatus.h:
* Source/JavaScriptCore/bytecode/PutByStatus.cpp:
(JSC::PutByStatus::PutByStatus):
(JSC::PutByStatus::computeFor):
* Source/JavaScriptCore/bytecode/Repatch.cpp:
(JSC::tryCacheGetBy):
(JSC::repatchGetBy):
(JSC::repatchGetBySlowPathCall):
(JSC::tryCachePutBy):
(JSC::tryCacheDeleteBy):
* Source/JavaScriptCore/bytecode/Repatch.h:
* Source/JavaScriptCore/bytecode/SetPrivateBrandStatus.cpp:
(JSC::SetPrivateBrandStatus::SetPrivateBrandStatus):
* Source/JavaScriptCore/bytecode/StructureStubInfo.cpp:
(JSC::StructureStubInfo::summary const):
* Source/JavaScriptCore/bytecode/StructureStubInfo.h:
(JSC::StructureStubInfo::considerCachingGeneric):
(JSC::StructureStubInfo::considerCachingBy):
(JSC::StructureStubInfo::considerCachingMegamorphic):
(JSC::StructureStubInfo::considerCachingImpl):
(JSC::StructureStubInfo::considerCaching): Deleted.
* Source/JavaScriptCore/bytecode/StubInfoSummary.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/bytecode/StubInfoSummary.h:
(JSC::isInlineable):
(JSC::slowVersion):
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::finalize):
* Source/JavaScriptCore/jit/JITOperations.cpp:
(JSC::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/jit/JITOperations.h:
* Source/JavaScriptCore/llint/LLIntSlowPaths.cpp:
(JSC::LLInt::LLINT_SLOW_PATH_DECL):
* Source/JavaScriptCore/runtime/CommonSlowPaths.cpp:
(JSC::JSC_DEFINE_COMMON_SLOW_PATH):
* Source/JavaScriptCore/runtime/HasOwnPropertyCache.h:
(JSC::VM::ensureHasOwnPropertyCache):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSObject.cpp:
(JSC::JSObject::setPrototypeDirect):
(JSC::JSObject::deleteProperty):
* Source/JavaScriptCore/runtime/JSObject.h:
* Source/JavaScriptCore/runtime/JSObjectInlines.h:
(JSC::JSObject::prepareToPutDirectWithoutTransition):
(JSC::JSObject::putDirectInternal):
(JSC::JSObject::mayBePrototype const):
(JSC::JSObject::didBecomePrototype):
* Source/JavaScriptCore/runtime/MegamorphicCache.cpp: Copied from Source/JavaScriptCore/bytecode/StubInfoSummary.cpp.
(JSC::MegamorphicCache::age):
(JSC::MegamorphicCache::clearEntries):
* Source/JavaScriptCore/runtime/MegamorphicCache.h: Added.
(JSC::MegamorphicCache::Entry::offsetOfUid):
(JSC::MegamorphicCache::Entry::offsetOfStructureID):
(JSC::MegamorphicCache::Entry::offsetOfEpoch):
(JSC::MegamorphicCache::Entry::offsetOfOffset):
(JSC::MegamorphicCache::Entry::offsetOfHolder):
(JSC::MegamorphicCache::Entry::initAsMiss):
(JSC::MegamorphicCache::Entry::initAsHit):
(JSC::MegamorphicCache::offsetOfPrimaryEntries):
(JSC::MegamorphicCache::offsetOfSecondaryEntries):
(JSC::MegamorphicCache::offsetOfEpoch):
(JSC::MegamorphicCache::primaryHash):
(JSC::MegamorphicCache::secondaryHash):
(JSC::MegamorphicCache::initAsMiss):
(JSC::MegamorphicCache::initAsHit):
(JSC::MegamorphicCache::epoch const):
(JSC::MegamorphicCache::bumpEpoch):
* Source/JavaScriptCore/runtime/ObjectPrototype.cpp:
(JSC::objectPrototypeHasOwnProperty):
* Source/JavaScriptCore/runtime/Structure.cpp:
(JSC::Structure::Structure):
(JSC::Structure::becomePrototypeTransition):
(JSC::Structure::nonPropertyTransitionSlow):
* Source/JavaScriptCore/runtime/Structure.h:
* Source/JavaScriptCore/runtime/StructureCache.cpp:
(JSC::StructureCache::createEmptyStructure):
* Source/JavaScriptCore/runtime/StructureInlines.h:
(JSC::Structure::create):
(JSC::Structure::nonPropertyTransition):
* Source/JavaScriptCore/runtime/StructureTransitionTable.h:
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::VM):
(JSC::VM::ensureMegamorphicCacheSlow):
(JSC::VM::invalidateStructureChainIntegrity):
* Source/JavaScriptCore/runtime/VM.h:
(JSC::VM::megamorphicCache):
(JSC::VM::ensureMegamorphicCache):
* Source/WebCore/bindings/js/JSDOMAsyncIterator.h:
(WebCore::JSDOMAsyncIteratorBase::createPrototype):
* Source/WebCore/bindings/js/JSDOMBuiltinConstructor.h:
(WebCore::JSDOMBuiltinConstructor&lt;JSClass&gt;::createStructure):
* Source/WebCore/bindings/js/JSDOMConstructor.h:
(WebCore::JSDOMConstructor&lt;JSClass&gt;::createStructure):
* Source/WebCore/bindings/js/JSDOMConstructorNotCallable.h:
(WebCore::JSDOMConstructorNotCallable&lt;JSClass&gt;::createStructure):
* Source/WebCore/bindings/js/JSDOMConstructorNotConstructable.h:
(WebCore::JSDOMConstructorNotConstructable&lt;JSClass&gt;::createStructure):
* Source/WebCore/bindings/js/JSDOMIterator.h:
(WebCore::JSDOMIteratorBase::createPrototype):
* Source/WebCore/bindings/js/JSWindowProxy.cpp:
(WebCore::JSWindowProxy::setWindow):
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(GenerateImplementation):
* Source/WebCore/workers/WorkerOrWorkletScriptController.cpp:
(WebCore::WorkerOrWorkletScriptController::initScriptWithSubclass):

Canonical link: <a href="https://commits.webkit.org/262808@main">https://commits.webkit.org/262808@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/999dc0ac8a471f6ba5c78048a43492e47de7af65

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2675 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2677 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2816 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4083 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3050 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/2638 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2818 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2773 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/2358 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2699 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3147 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2392 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/3846 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/640 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2377 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/2241 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/2228 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/2759 "Built successfully") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2393 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/3605 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/2557 "Built successfully and passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2428 "Passed tests") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/2211 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/2752 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2430 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2380 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/646 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2418 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/2812 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/312 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2571 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/761 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
```diff
@@ -4859,10 +4859,14 @@ sub GenerateImplementation
             my $parentClassNameForPrototype = "JS" . $interface->parentType->name;
-            push(@implContent, "    return ${className}Prototype::create(vm, &globalObject, ${className}Prototype::createStructure(vm, &globalObject, ${parentClassNameForPrototype}::prototype(vm, globalObject)));\n");
+            push(@implContent, "    auto* structure = ${className}Prototype::createStructure(vm, &globalObject, ${parentClassNameForPrototype}::prototype(vm, globalObject));\n");
+            push(@implContent, "    structure->setMayBePrototype(true);\n");
```
If we accidentally forget to set this, it just results in a perf hit right?

justinmichaud replied:
It would be nice if we had a more ergonomic way to set this

Constellation replied:
If we forget this, basically it does not affect things much. `didBecomePrototype` automatically does this transition. This is kind of optimization that we put this to structures a-priori if it is *definitely* a prototoype's Structure. It reduces Structure transition counts. But not affecting on benchmark score.

justinmichaud commented on Source/JavaScriptCore/runtime/StructureTransitionTable.h:
```diff
@@ -37,28 +37,30 @@ class Structure;
 
 using TransitionPropertyAttributes = uint8_t;
 
+// This must be 5 bits (less than 32).
```
static assert?

Constellation replied:
Not much possible since there is no way to get the max entry of TransitionKind.

justinmichaud commented on Source/JavaScriptCore/bytecode/PutByStatus.cpp:
```diff
@@ -394,6 +395,10 @@ PutByStatus PutByStatus::computeFor(JSGlobalObject* globalObject, const Structur
+
+        // If the structure is for prototype, we should do a slow path which can invalidate MegamorphicCache.
+        if (structure->mayBePrototype())
+            return PutByStatus(LikelyTakesSlowPath);
```
Hmm, so previously we could have had a prototype set be cached + inlined, but not anymore? I guess this happens pretty infrequently

Constellation replied:
This happens very infrequently, and it didn't cause any performance issues (I isolated this part in a separate A/B test and ensured).

justinmichaud commented on Source/JavaScriptCore/bytecode/StructureStubInfo.h:
```diff
@@ -268,6 +278,9 @@ class StructureStubInfo {
             }
             
             bufferingCountdown--;
+
```
So what is an example of a non-cell that the megamorphic cache would want to cache? A property miss on null?

Constellation replied:
`considerCachingMegamorphic` is used for megamoprhic cache's miss. Non cell case, we eventually go to generic case.

justinmichaud commented on Source/JavaScriptCore/bytecode/StubInfoSummary.h:
```diff
@@ -39,6 +40,7 @@ inline bool isInlineable(StubInfoSummary summary)
 {
     switch (summary) {
     case StubInfoSummary::Simple:
+    case StubInfoSummary::Megamorphic:
```
GetByStatus can't inline this in DFG, right?

Constellation replied:
Yes. We try using IC again to see / collect data in DFG / FTL context.

justinmichaud commented on Source/JavaScriptCore/runtime/MegamorphicCache.h:
```diff
@@ -0,0 +1,150 @@
+
+    ALWAYS_INLINE static uint32_t primaryHash(StructureID structureID, UniquedStringImpl* uid)
+    {
+        uint32_t sid = bitwise_cast<uint32_t>(structureID);
```
It would be interesting to collect a list of StructureIDs and measure how uniform these hash functions actually are

Constellation replied:
Currently, we made it work well for various benchmark cases. And so far, primary / secondary hash can cover 95~% cases.

justinmichaud commented on Source/JavaScriptCore/bytecode/Repatch.cpp:
```diff
@@ -596,17 +597,29 @@ static InlineCacheAction tryCacheGetBy(JSGlobalObject* globalObject, CodeBlock*
+{
+    resetGetBy(codeBlock, stubInfo, kind);
+    repatchSlowPathCall(codeBlock, stubInfo, appropriateGetByFunction(kind));
+    // dataLogLn("GENERIC");
```
nit

Constellation replied:
Removed

justinmichaud commented on Source/JavaScriptCore/jit/JITOperations.cpp:
```diff
@@ -347,6 +348,68 @@ JSC_DEFINE_JIT_OPERATION(operationGetByIdDirectOptimize, EncodedJSValue, (JSGlob
+    while (true) {
+        if (UNLIKELY(TypeInfo::overridesGetOwnPropertySlot(object->inlineTypeFlags()) && object->type() != ArrayType && object->type() != JSFunctionType && object != globalObject->arrayPrototype())) {
+            if (stubInfo->considerCachingMegamorphic(vm))
+                repatchGetBySlowPathCall(callFrame->codeBlock(), *stubInfo, GetByKind::ById);
```
So we are already repatched to be megamorphic, so this case will make us generic? Why don't we always become generic in this case?

justinmichaud replied:
Oh, I see. This function makes the megamorphic cache, and so therefore the slow path needs to become the generic slow path. If the megamorphic cache fails, where do we completely abort?

Constellation replied:
This is megamorphic cache's slow path. This code can handle any property access. And when we found this is not cacheable, we record it in stubInfo, and if we see it frequently, we disable megamorphic cache for this callsite.

justinmichaud replied:
So why do we only disable the megamorphic cache if considerCachingMegamorphic? Isn't that backwards?

Constellation replied:
No, this is correct. considerCaching thing (cache re-considering countdown) for Megamorphic calls.

justinmichaud commented on Source/JavaScriptCore/jit/JITOperations.cpp:
```diff
@@ -347,6 +348,68 @@ JSC_DEFINE_JIT_OPERATION(operationGetByIdDirectOptimize, EncodedJSValue, (JSGlob
+            if (LIKELY(cacheable && slot.isCacheableValue() && slot.cachedOffset() <= MegamorphicCache::maxOffset))
+                vm.megamorphicCache()->initAsHit(baseObject->structureID(), uid, slot.slotBase(), slot.cachedOffset(), slot.slotBase() == baseObject);
+            else {
+                if (stubInfo->considerCachingMegamorphic(vm))
```
ditto, I don't know why we check this before becoming generic

justinmichaud replied:
nvm

justinmichaud commented on Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:
```diff
@@ -1258,6 +1260,90 @@ void InlineCacheCompiler::generateWithGuard(AccessCase& accessCase, CCallHelpers
+        jit.xor32(scratch2GPR, scratch3GPR);
+        jit.add32(CCallHelpers::TrustedImm32(hash), scratch3GPR);
+        jit.and32(CCallHelpers::TrustedImm32(MegamorphicCache::primaryMask), scratch3GPR);
+        if (hasOneBitSet(sizeof(MegamorphicCache::Entry))) // is a power of 2
```
Can we just static_assert the fast thing?

Constellation replied:
We can make it `if-constexpr`, but not static_assert.

justinmichaud commented on Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:
```diff
@@ -1258,6 +1260,90 @@ void InlineCacheCompiler::generateWithGuard(AccessCase& accessCase, CCallHelpers
+            jit.mul32(CCallHelpers::TrustedImm32(sizeof(MegamorphicCache::Entry)), scratch3GPR, scratch3GPR);
+        auto& cache = vm.ensureMegamorphicCache();
+        jit.move(CCallHelpers::TrustedImmPtr(&cache), scratch2GPR);
+        ASSERT(!MegamorphicCache::offsetOfPrimaryEntries());
```
static_assert?

Constellation replied:
It is not possible since OBJECT_OFFSETOF is not constexpr.

justinmichaud commented on Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:
```diff
@@ -1258,6 +1260,90 @@ void InlineCacheCompiler::generateWithGuard(AccessCase& accessCase, CCallHelpers
+        jit.urshift32(scratch2GPR, CCallHelpers::TrustedImm32(MegamorphicCache::structureIDHashShift3), scratch3GPR);
+        jit.add32(scratch2GPR, scratch3GPR);
+        jit.and32(CCallHelpers::TrustedImm32(MegamorphicCache::secondaryMask), scratch3GPR);
+        if (hasOneBitSet(sizeof(MegamorphicCache::Entry))) // is a power of 2
```
ditto

Constellation replied:
Ditto.

justinmichaud commented on Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:
```diff
@@ -2598,6 +2685,54 @@ AccessGenerationResult InlineCacheCompiler::regenerate(const GCSafeConcurrentJSL
+        case AccessType::GetById: {
+            auto identifier = cases.last()->m_identifier;
+            bool allAreSimpleLoadOrMiss = true;
+            for (auto& accessCase : cases) {
```
So if one of the cases is strange, but the other 32 are cacheable, we won't turn this call site into a megamorphic IC?

justinmichaud replied:
And none of these cases make their way into the final global cache, right? It has to go through the megamorphic slow path?

Constellation replied:
These structures are captured via sampling, so, if they include something different, then we should assume that we will see this different structure fair amount of rate. So, I think just giving up to the old path is the right path.

justinmichaud commented on Source/JavaScriptCore/bytecode/StructureStubInfo.h:
```diff
@@ -204,28 +204,38 @@ class StructureStubInfo {
     }
 private:
-    ALWAYS_INLINE bool considerCaching(VM& vm, CodeBlock* codeBlock, Structure* structure, CacheableIdentifier impl)
+    ALWAYS_INLINE bool considerCachingImpl(VM& vm, CodeBlock* codeBlock, Structure* structure, CacheableIdentifier impl)
```
Is the intent of this function to decide to repatch? Would it be correct to rename it to considerChangingCache or considerRepatchingCache? In the megamorphic case, we should not be adding new access cases, so this is really only making sure that we don't remove the megamorphic cache on the first miss, right?

Constellation replied:
Yes, this function is used to decide the repatching, which spmetimes making IC as generic (as before. Previously, we did make the IC generic if we see too many cases / see GiveUp case).
I think renaming is OK

justinmichaud reviewed (approved):
rs=me

MenloDorian commented on Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:
```diff
@@ -49,7 +49,9 @@
 #include "LinkBuffer.h"
+#include "MegamorphicCache.h"
 #include "ModuleNamespaceAccessCase.h"
+#include "ProbeContext.h"
```
This is not needed, right?

Constellation replied:
Yes, this is not necessary.

Constellation committed:
[JSC] Add megamorphic cache
https://bugs.webkit.org/show_bug.cgi?id=255060
rdar://problem/107683366

Reviewed by Justin Michaud.

This patch adds megamorphic load cache finally, which exists in SpiderMonkey and V8. Our plan is adding megamorphic store cache next.

1. We move mayBePrototype bit from object cell to Structure. This allows us to disable IC setup for prototype objects by observing Structure.
   (And this is extremely rare). And every time we do some modification of the object, we can check `mayBePrototype()` and we can invalidate
   megamorphic cache integrity (VM::invalidateStructureChainIntegrity). This bit & transition happens automatically when an object gets used as a prototype.
   But we mark all DOM prototypes and constructors a priori since they become prototype objects.
2. We integrate megamorphic load cache, which is 2-level hash table with 2 different hash functions. Primary table has 2048 entries and secondaryHash
   table has 512 entries, which is derived number from V8's megamorphic cache implementation. Our megamorphic cache's entry is StructureID + UniquedStringImpl.
   And the result is holder object and offset. Holder is null if property is missed. And holder is special marker (0x2) if it is own property.
   Every time objects used for prototype gets changed, whole megamorphic cache gets invalidated by incrementing epoch. This is surprisingly rare since
   we do not modify objects used for prototype after creating instances (And once the page gets stable state, we do not modify prototype objects).
   We use 2-level cache with 2 completely different hash functions to make it work well for various StructureID + UniquedStringImpl* pair.
3. We increment epoch of megamorphic cache when GC happens. So we do not keep entries valid over GC.

This change roughly offers 2-3x performance improvement for megamorphic access.

                                           ToT                     Patched

    megamorphic-own-load             99.3922+-1.9842     ^     41.2465+-0.0893        ^ definitely 2.4097x faster
    megamorphic-load                 14.8899+-0.0635     ^      5.0489+-0.0328        ^ definitely 2.9491x faster
    megamorphic-prototype-load      136.9894+-0.6358     ^     38.1426+-0.0790        ^ definitely 3.5915x faster
    megamorphic-miss                 98.3712+-1.6652     ^     31.9041+-0.0640        ^ definitely 3.0833x faster

* JSTests/microbenchmarks/megamorphic-miss.js: Added.
(f):
* JSTests/microbenchmarks/megamorphic-own-load.js: Added.
(f):
* JSTests/microbenchmarks/megamorphic-prototype-load.js: Added.
(f):
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/bytecode/AccessCase.cpp:
(JSC::AccessCase::create):
(JSC::AccessCase::guardedByStructureCheckSkippingConstantIdentifierCheck const):
(JSC::AccessCase::requiresIdentifierNameMatch const):
(JSC::AccessCase::requiresInt32PropertyCheck const):
(JSC::AccessCase::needsScratchFPR const):
(JSC::AccessCase::forEachDependentCell const):
(JSC::AccessCase::doesCalls const):
(JSC::AccessCase::canReplace const):
(JSC::AccessCase::runWithDowncast):
(JSC::AccessCase::canBeShared):
* Source/JavaScriptCore/bytecode/AccessCase.h:
* Source/JavaScriptCore/bytecode/CheckPrivateBrandStatus.cpp:
(JSC::CheckPrivateBrandStatus::CheckPrivateBrandStatus):
* Source/JavaScriptCore/bytecode/DeleteByStatus.cpp:
(JSC::DeleteByStatus::DeleteByStatus):
* Source/JavaScriptCore/bytecode/GetByStatus.cpp:
(JSC::GetByStatus::GetByStatus):
* Source/JavaScriptCore/bytecode/InByStatus.h:
* Source/JavaScriptCore/bytecode/InlineCacheCompiler.cpp:
(JSC::InlineCacheCompiler::generateWithGuard):
(JSC::InlineCacheCompiler::generateImpl):
(JSC::InlineCacheCompiler::regenerate):
(WTF::printInternal):
* Source/JavaScriptCore/bytecode/InlineCacheCompiler.h:
(JSC::AccessGenerationResult::AccessGenerationResult):
(JSC::AccessGenerationResult::generatedMegamorphicCode const):
(JSC::AccessGenerationResult::generatedSomeCode const):
* Source/JavaScriptCore/bytecode/InstanceOfStatus.h:
* Source/JavaScriptCore/bytecode/PutByStatus.cpp:
(JSC::PutByStatus::PutByStatus):
(JSC::PutByStatus::computeFor):
* Source/JavaScriptCore/bytecode/Repatch.cpp:
(JSC::tryCacheGetBy):
(JSC::repatchGetBy):
(JSC::repatchGetBySlowPathCall):
(JSC::tryCachePutBy):
(JSC::tryCacheDeleteBy):
* Source/JavaScriptCore/bytecode/Repatch.h:
* Source/JavaScriptCore/bytecode/SetPrivateBrandStatus.cpp:
(JSC::SetPrivateBrandStatus::SetPrivateBrandStatus):
* Source/JavaScriptCore/bytecode/StructureStubInfo.cpp:
(JSC::StructureStubInfo::summary const):
* Source/JavaScriptCore/bytecode/StructureStubInfo.h:
(JSC::StructureStubInfo::considerCachingGeneric):
(JSC::StructureStubInfo::considerCachingBy):
(JSC::StructureStubInfo::considerCachingMegamorphic):
(JSC::StructureStubInfo::considerCachingImpl):
(JSC::StructureStubInfo::considerCaching): Deleted.
* Source/JavaScriptCore/bytecode/StubInfoSummary.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/bytecode/StubInfoSummary.h:
(JSC::isInlineable):
(JSC::slowVersion):
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/heap/Heap.cpp:
(JSC::Heap::finalize):
* Source/JavaScriptCore/jit/JITOperations.cpp:
(JSC::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/jit/JITOperations.h:
* Source/JavaScriptCore/llint/LLIntSlowPaths.cpp:
(JSC::LLInt::LLINT_SLOW_PATH_DECL):
* Source/JavaScriptCore/runtime/CommonSlowPaths.cpp:
(JSC::JSC_DEFINE_COMMON_SLOW_PATH):
* Source/JavaScriptCore/runtime/HasOwnPropertyCache.h:
(JSC::VM::ensureHasOwnPropertyCache):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSObject.cpp:
(JSC::JSObject::setPrototypeDirect):
(JSC::JSObject::deleteProperty):
* Source/JavaScriptCore/runtime/JSObject.h:
* Source/JavaScriptCore/runtime/JSObjectInlines.h:
(JSC::JSObject::prepareToPutDirectWithoutTransition):
(JSC::JSObject::putDirectInternal):
(JSC::JSObject::mayBePrototype const):
(JSC::JSObject::didBecomePrototype):
* Source/JavaScriptCore/runtime/MegamorphicCache.cpp: Copied from Source/JavaScriptCore/bytecode/StubInfoSummary.cpp.
(JSC::MegamorphicCache::age):
(JSC::MegamorphicCache::clearEntries):
* Source/JavaScriptCore/runtime/MegamorphicCache.h: Added.
(JSC::MegamorphicCache::Entry::offsetOfUid):
(JSC::MegamorphicCache::Entry::offsetOfStructureID):
(JSC::MegamorphicCache::Entry::offsetOfEpoch):
(JSC::MegamorphicCache::Entry::offsetOfOffset):
(JSC::MegamorphicCache::Entry::offsetOfHolder):
(JSC::MegamorphicCache::Entry::initAsMiss):
(JSC::MegamorphicCache::Entry::initAsHit):
(JSC::MegamorphicCache::offsetOfPrimaryEntries):
(JSC::MegamorphicCache::offsetOfSecondaryEntries):
(JSC::MegamorphicCache::offsetOfEpoch):
(JSC::MegamorphicCache::primaryHash):
(JSC::MegamorphicCache::secondaryHash):
(JSC::MegamorphicCache::initAsMiss):
(JSC::MegamorphicCache::initAsHit):
(JSC::MegamorphicCache::epoch const):
(JSC::MegamorphicCache::bumpEpoch):
* Source/JavaScriptCore/runtime/ObjectPrototype.cpp:
(JSC::objectPrototypeHasOwnProperty):
* Source/JavaScriptCore/runtime/Structure.cpp:
(JSC::Structure::Structure):
(JSC::Structure::becomePrototypeTransition):
(JSC::Structure::nonPropertyTransitionSlow):
* Source/JavaScriptCore/runtime/Structure.h:
* Source/JavaScriptCore/runtime/StructureCache.cpp:
(JSC::StructureCache::createEmptyStructure):
* Source/JavaScriptCore/runtime/StructureInlines.h:
(JSC::Structure::create):
(JSC::Structure::nonPropertyTransition):
* Source/JavaScriptCore/runtime/StructureTransitionTable.h:
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::VM):
(JSC::VM::ensureMegamorphicCacheSlow):
(JSC::VM::invalidateStructureChainIntegrity):
* Source/JavaScriptCore/runtime/VM.h:
(JSC::VM::megamorphicCache):
(JSC::VM::ensureMegamorphicCache):
* Source/WebCore/bindings/js/JSDOMAsyncIterator.h:
(WebCore::JSDOMAsyncIteratorBase::createPrototype):
* Source/WebCore/bindings/js/JSDOMBuiltinConstructor.h:
(WebCore::JSDOMBuiltinConstructor<JSClass>::createStructure):
* Source/WebCore/bindings/js/JSDOMConstructor.h:
(WebCore::JSDOMConstructor<JSClass>::createStructure):
* Source/WebCore/bindings/js/JSDOMConstructorNotCallable.h:
(WebCore::JSDOMConstructorNotCallable<JSClass>::createStructure):
* Source/WebCore/bindings/js/JSDOMConstructorNotConstructable.h:
(WebCore::JSDOMConstructorNotConstructable<JSClass>::createStructure):
* Source/WebCore/bindings/js/JSDOMIterator.h:
(WebCore::JSDOMIteratorBase::createPrototype):
* Source/WebCore/bindings/js/JSWindowProxy.cpp:
(WebCore::JSWindowProxy::setWindow):
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(GenerateImplementation):
* Source/WebCore/workers/WorkerOrWorkletScriptController.cpp:
(WebCore::WorkerOrWorkletScriptController::initScriptWithSubclass):

Canonical link: https://commits.webkit.org/262808@main
